agent/fw_cakephp.c

```diff
@@ -281,35 +268,38 @@ NR_PHP_WRAPPER(nr_cakephp_problem_1) {
 NR_PHP_WRAPPER_END
 
 /*
- * CakePHP 2.0+
+ * CakePHP 4.0+
  *
- * If the action or controller is not found during the dispatch process, the
- * appropriate Exception will be created and thrown. We wrap the CakeException
- * constructor instead of the Exception handler, since CakePHP allows for the
- * handler to be completely replaced.
- *
- *  CakePHP 2.x is end-of-life and in maintenance mode (critical bugfixes only).
- * As such, functionality added in PHP 7.1+ is not well supported.
- *
- * txn naming scheme:
- * In this case, `nr_txn_set_path` is called before `NR_PHP_WRAPPER_CALL` with
- * `NR_NOT_OK_TO_OVERWRITE` and as this corresponds to calling the wrapped
- * function in func_begin it needs to be explicitly set as a before_callback to
- * ensure OAPI compatibility. This entails that the first wrapped call gets to
- * name the txn.
+ * Report errors and exceptions caught by CakePHP's error handler.
  */
-NR_PHP_WRAPPER(nr_cakephp_problem_2) {
-  const char* name = "Exception";
+NR_PHP_WRAPPER(nr_cakephp_error_handler_wrapper) {
+  zval* exception = NULL;
 
-  (void)wraprec;
   NR_UNUSED_SPECIALFN;
+  (void)wraprec;
 
   NR_PHP_WRAPPER_REQUIRE_FRAMEWORK(NR_FW_CAKEPHP);
 
-  nr_txn_set_path("CakePHP", NRPRG(txn), name, NR_PATH_TYPE_ACTION,
+  exception = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS);
+  if (!nr_php_is_zval_valid_object(exception)) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "%s: exception is NULL or not an object",
+                     __func__);
+    goto end;
+  }
+
+  if (NR_SUCCESS
+      != nr_php_error_record_exception(
+          NRPRG(txn), exception, nr_php_error_get_priority(E_ERROR), true,
+          "Uncaught exception ", &NRPRG(exception_filters))) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "%s: unable to record exception", __func__);
+  }
+
+  nr_txn_set_path("CakePHP", NRPRG(txn),
+                  nr_php_get_server_global("REQUEST_URI"), NR_PATH_TYPE_ACTION,
```

Comment:
This will result in memory leak: `nr_php_get_server_global` calls `nr_strndup` and result of `nr_strndup` is returned to the caller and the caller is responsible for freeing the memory. Look [here](https://github.com/newrelic/newrelic-php-agent/blob/R11.3/agent/php_txn.c#L46-L52) for an example of a proper use of `nr_php_get_server_global`.

Reply:
ah, so previously (https://github.com/newrelic/newrelic-php-agent/pull/983#discussion_r1838710212) it was being dup'ed twice

Reply:
> ah, so previously ([#983 (comment)](https://github.com/newrelic/newrelic-php-agent/pull/983#discussion_r1838710212)) it was being dup'ed twice

Yes.

Reply:
Good catch, fixed: [2a8975e](https://github.com/newrelic/newrelic-php-agent/pull/983/commits/2a8975e2df808c464a50ae6ffa3266e9cebbbdba)

Reply:
[2a8975e](https://github.com/newrelic/newrelic-php-agent/pull/983/commits/2a8975e2df808c464a50ae6ffa3266e9cebbbdba) is not fixing the memory leak - the return value from `nr_php_get_server_global` needs to be freed by the caller. Look [here](https://github.com/newrelic/newrelic-php-agent/blob/R11.3/agent/php_txn.c#L46-L52) for an example of a proper use of nr_php_get_server_global.

Reply:
Thanks for catching that, I didn't need to do another `nr_strdup`. Fixed: [fd5ef92](https://github.com/newrelic/newrelic-php-agent/pull/983/commits/fd5ef92a51c39cd4f1d6bfc4fb9b718df237e292)